### PR TITLE
Workflow on CRD

### PR DESCRIPTION
**Pull request by sdminonne**
1st commit contains removing old vendoring and  all new code
2dn commit just vendoring
@alicealbano1A @ericmountain-1A @clamoriniere1A @dbenque-1a

### CODE REVIEW

File: pkg/api/v1/default.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+			Spec: step.JobTemplate.Spec,
+		}
+		SetDefaults_Job(dummyJob)
+		step.JobTemplate.Spec = dummyJob.Spec
+		w.Spec.Steps[i] = step
+	}
+	return w
+}
+
+// SetDefaults_Job copied here function from k8s.io/kubernetes/pkg/apis/batch/v1/defaults.go
```
[the remark this answers is not in the document]
Comment (clamoriniere1A) [reply pkg/api/v1/default.go]: You should add in the comment the version used for this copy, and also in the `README.md` that you need to update this file when you re-vendor the kubernetes client.

File: pkg/api/v1/util.go [hunk 2]
```diff
@@ -1,9 +1,9 @@
-package api
+package v1
 
 import "fmt"
 
 // RemoveStepFromSpec remove Step from Workflow Spec
```
Comment (clamoriniere1A) [inline pkg/api/v1/util.go]: s/remove/removes

File: pkg/api/v1/validation.go [hunk 3]
```diff
@@ -1,69 +1,47 @@
...
-Unless required by applicable law or agreed to in writing, software
-distributed under the License is distributed on an "AS IS" BASIS,
-WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-See the License for the specific language governing permissions and
-limitations under the License.
-*/
+import (
 
-package validation
+	//"k8s.io/apimachinery/pkg/util/runtime"
```
Comment (clamoriniere1A) [inline pkg/api/v1/validation.go]: dead code

File: pkg/api/v1/validation.go [hunk 4]
```diff
@@ -1,69 +1,47 @@
...
-func ValidateWorkflowSpec(spec *wapi.WorkflowSpec, fieldPath *field.Path) field.ErrorList {
+// ValidateWorkflowSpec validates WorkflowSpec
+func ValidateWorkflowSpec(spec *WorkflowSpec, fieldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
 	if spec.ActiveDeadlineSeconds != nil {
-		allErrs = append(allErrs, kapivalidation.ValidateNonnegativeField(int64(*spec.ActiveDeadlineSeconds), fieldPath.Child("activeDeadlineSeconds"))...)
+		allErrs = append(allErrs, validation.ValidateNonnegativeField(int64(*spec.ActiveDeadlineSeconds), fieldPath.Child("activeDeadlineSeconds"))...)
 	}
 	if spec.Selector == nil {
 		allErrs = append(allErrs, field.Required(fieldPath.Child("selector"), ""))
```
Comment (clamoriniere1A) [inline pkg/api/v1/validation.go]: maybe you can use `const string`for "selector", "steps", "spec"...
it is used in several places.

File: pkg/api/v1/validation.go [hunk 5]
```diff
@@ -252,8 +220,56 @@ func ValidateWorkflowSpecUpdate(spec, oldSpec *wapi.WorkflowSpec, running, compl
 	return allErrs
 }
 
-func ValidateWorkflowStatusUpdate(status, oldStatus wapi.WorkflowStatus) field.ErrorList {
+// ValidateWorkflowStatus validates status
+func ValidateWorkflowStatus(status *WorkflowStatus, fieldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	//glog.V(6).Infof("Validating WorkflowStatus %v\n", spew.Sdump(*status))
```
Comment (clamoriniere1A) [inline pkg/api/v1/validation.go]: dead code

File: pkg/controller/job_control.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+			Labels:       desiredLabels,
+			Annotations:  desiredAnnotations,
+			GenerateName: jobGeneratedName, // TODO: double check prefix
+		},
+	}
+	job.Spec = *template.Spec.DeepCopy()
+	return job, nil
+}
+
+// FakeJobControl implementes WorkflowJobControl interface for testing purpouse
```
Comment (clamoriniere1A) [inline pkg/controller/job_control.go]: s/implementes/implements
s/purpouse/purpose

File: pkg/api/v1/default.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sdminonne) [reply pkg/api/v1/default.go]: Right on both.

File: pkg/controller/job_control.go (hunk 6 above)
Comment (sdminonne) [reply pkg/controller/job_control.go]: thanks

File: pkg/api/v1/util.go (hunk 2 above)
Comment (sdminonne) [reply pkg/api/v1/util.go]: thanks

File: pkg/api/v1/validation.go (hunk 3 above)
Comment (sdminonne) [reply pkg/api/v1/validation.go]: thanks

File: pkg/api/v1/validation.go (hunk 5 above)
Comment (sdminonne) [reply pkg/api/v1/validation.go]: right

File: pkg/api/v1/validation.go (hunk 4 above)
Comment (sdminonne) [reply pkg/api/v1/validation.go]: ok

File: app/config.go [hunk 7]
```diff
@@ -67,8 +33,4 @@ func NewWorkflowControllerConfig() *Config {
 // AddFlags add cobra flags to populate Config
 func (c *Config) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&c.KubeConfigFile, "kubeconfig", c.KubeConfigFile, "Location of kubecfg file for access to kubernetes master service; --kube-master-url overrides the URL part of this; if neither this nor --kube-master-url are provided, defaults to service account tokens")
```
[the remark this answers is not in the document]
Comment (sdminonne) [reply app/config.go]: cruft, right

File: pkg/client/client.go [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+		}
+		for _, cond := range crd.Status.Conditions {
+			switch cond.Type {
+			case apiextensionsv1beta1.Established:
+				if cond.Status == apiextensionsv1beta1.ConditionTrue {
+					return true, err
+				}
+			case apiextensionsv1beta1.NamesAccepted:
+				if cond.Status == apiextensionsv1beta1.ConditionFalse {
+					fmt.Printf("Name conflict: %v\n", cond.Reason)
```
[the remark this answers is not in the document]
Comment (sdminonne) [reply pkg/client/client.go]: right. need glog

File: pkg/controller/controller.go [hunk 9]
```diff
@@ -0,0 +1,613 @@
...
+	JobLister      batchv1listers.JobLister
+	JobStoreSynced cache.InformerSynced // returns true if job stroe has been synced. Added as member for testing
+
+	JobControl JobControlInterface
+
+	updateHandler func(*wapi.Workflow) error // callback to upate Workflow. Added as member for testing
+
+	queue workqueue.RateLimitingInterface // Workflows to be synced
+
+	cleanQueue workqueue.RateLimitingInterface // Workflows to be removed
```
[the remark this answers is not in the document]
Comment (sdminonne) [reply pkg/controller/controller.go]: going to remove cleanQueue

File: pkg/controller/controller.go [hunk 10]
```diff
@@ -0,0 +1,613 @@
...
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    wc.onAddWorkflow,
+			UpdateFunc: wc.onUpdateWorkflow,
+			DeleteFunc: wc.onDeleteWorkflow,
+		},
+	)
+
+	wc.JobControl = &WorkflowJobControl{kubeClient, wc.Recorder}
+
+	wc.JobInformer = cache.NewSharedIndexInformer(
```
[the remark this answers is not in the document]
Comment (sdminonne) [reply pkg/controller/controller.go]: Going to replace with the factory. Thanks

File: pkg/controller/controller.go [hunk 11]
```diff
@@ -0,0 +1,613 @@
...
+	defer func() {
+		glog.V(6).Infof("Finished syncing Workflow %q (%v", key, time.Now().Sub(startTime))
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	glog.V(6).Infof("Syncing %s/%s", namespace, name)
+
```
[the remark this answers is not in the document]
Comment (sdminonne) [reply pkg/controller/controller.go]: going to remove the cleanQueue. Thanks

File: pkg/controller/controller.go [hunk 12]
```diff
@@ -0,0 +1,613 @@
...
+		glog.Errorf("unable to get Workflow %s/%s:%v", namespace, name, err)
+		return nil
+	}
+	if !exists {
+		glog.V(6).Info("unable to get Workflow %s/%s: %v (may be deleted?)", namespace, name)
+		return nil
+	}
+	sharedWorkflow := obj.(*wapi.Workflow)
+	// Defaulting...
+	if !sharedWorkflow.Status.Validated && // a Workflow validated has been already defaulted
```
[the remark this answers is not in the document]
Comment (sdminonne) [reply pkg/controller/controller.go]: I agree. thanks for this

File: pkg/api/v1/default.go [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+		obj.Spec.Parallelism = new(int32)
+		*obj.Spec.Parallelism = 1
+	}
+	if obj.Spec.Parallelism == nil {
+		obj.Spec.Parallelism = new(int32)
+		*obj.Spec.Parallelism = 1
+	}
+	labels := obj.Spec.Template.Labels
+	if labels != nil && len(obj.Labels) == 0 {
+		obj.Labels = labels
```
[the remark this answers is not in the document]
Comment (sdminonne) [reply pkg/api/v1/default.go]: Hmm I missed this. Don't remember the original reason. I think there's none.
Going to remove it. Thanks

**Comment by sdminonne** [comment]
@clamoriniere1A @dbenque-1a thanks folks.
Going to update with your suggestion.

File: pkg/api/v1/default.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (sdminonne) [reply pkg/api/v1/default.go]: It's the way it must work the defaulting mechanism.  Link between workflow and jobs is not done here.

File: pkg/controller/controller.go [hunk 14]
```diff
@@ -0,0 +1,613 @@
...
+		workflow.Status.Validated = true
+		if err := w.updateHandler(workflow); err != nil {
+			glog.Errorf("WorkflowController.sync Workflow %s/%s unable to set status.Validated to true:%v", namespace, name, err)
+			return fmt.Errorf("unable to set status.Validated to true:%s/%s:%v", namespace, name, err)
+		}
+		return nil
+	}
+
+	// Init status.StartTime
+	if workflow.Status.StartTime == nil {
```
[the remark this answers is not in the document]
Comment (sdminonne) [reply pkg/controller/controller.go]: I disagree. We must be sure to advance step by step. A resource must start to have a status (with a start time) when it's well-formed (defaulted and valid).

File: pkg/api/v1/validation.go (hunk 4 above)
Comment (sdminonne) [reply pkg/api/v1/validation.go]: Hmm taking back. Following kube convention I'll keep the strings without const.

Comment (clamoriniere1A) [reply pkg/api/v1/validation.go]: kube: the source of truth ;)

File: docs/devel.md [hunk 15]
```diff
@@ -0,0 +1,13 @@
+# How to generate deep-copy
+
+First of all you need `deepcopy-gen`. You can get directy from `gengo` package via a simple:
```
Comment (EricMountain-1A) [inline docs/devel.md]: s/get directy/retrieve directly/

File: examples/hello_workflow/workflow.yaml [hunk 16]
```diff
@@ -1,5 +1,5 @@
+apiVersion: "dag.example.com/v1"
```
Comment (EricMountain-1A) [inline examples/hello_workflow/workflow.yaml]: Shouldn't this be something other than "example.com".  I don't know... dag.acs.amadeus.com?

**Comment by sdminonne** [comment]
@EricMountain-1A thanks. Modifying right now.

File: examples/hello_workflow/workflow.yaml (hunk 16 above)
Comment (sdminonne) [reply examples/hello_workflow/workflow.yaml]: Right

File: docs/devel.md (hunk 15 above)
Comment (sdminonne) [reply docs/devel.md]: thanks

File: examples/hello_workflow/workflow.yaml (hunk 16 above)
Comment (sdminonne) [reply examples/hello_workflow/workflow.yaml]: It's going to be something configurable. But I'm going to do in another PR. Thanks

**Pull request merged**